attribution/api_attribution.py

```diff
@@ -46,11 +52,244 @@ def __init__(
+                            self.tokenizer.decode(perturbation_strategy.get_replacement_token(token_id)).strip()
+                            for token_id in ids_per_unit[i]
+                        ]
+                        perturbed_input.append(perturbed_unit)
+                        perturbed_units.append(unit)
+                    else:
+                        perturbed_input.append([unit])
+
+                perturbation = PerturbedLLMInput(
+                    input_units=perturbed_input, 
+                    masked_units=perturbed_units, 
+                    unit_idx=np.where(mask)[0].tolist(),
+                    tokenizer=self.tokenizer,
+                )
+
+                perturbations.append(perturbation)
+            
+            if verbose:
+                print("Masked out tokens/words:")
+                print(*[[perturbation.masked_string] for perturbation in perturbations], sep="\n")
+            
+            outputs = await self.get_multiple_completions([perturbation.input_string for perturbation in perturbations])
+            total_llm_calls += len(outputs)
+            
+            chunk_scores = []
+            unit_attribution = np.full((len(masks), unit_count), np.nan)
+            
+            for i, (perturbation, output, mask) in enumerate(zip(perturbations, outputs, masks)):
+                
+                for strategy in attribution_strategies:
+                    attribution_scores, norm_attribution_scores = self._get_scores(
+                        perturbed_output=output, 
+                        original_output=original_output, 
+                        attribution_strategy=strategy,
+                        chunksize=sum(mask),
+                        ignore_output_token_location=ignore_output_token_location,
+                    )
+
+                    if logger:
+                        logger.log_attributions(
+                            perturbation,
+                            norm_attribution_scores,
+                            strategy,
+                            output.message.content,
+                            depth=stage,
+                        )
+
+                        logger.log_perturbation(
+                            len(perturbation.masked_units) - 1,
+                            perturbation.masked_string,
+                            str(perturbation_strategy),
+                            original_input,
+                            original_output.message.content,
+                            perturbation.input_string,
+                            output.message.content,
+                        )
+                
+                    # For scoring we only use the first attribution strategy
+                    if strategy == attribution_strategies[0]:
+                        chunk_scores.append(attribution_scores["sentence_attribution"])
+                        unit_attribution[i, mask] = norm_attribution_scores["sentence_attribution"]
```

Comment:
Maybe this can be a separate function as well. It's similar to the code in `compute_attributions`, I wonder if we can reuse a single function.

Reply:
Moved logging inside `_get_scores` method